File: 532-k-diff-pairs-in-an-array/532-k-diff-pairs-in-an-array.java
class Solution {
    public int findPairs(int[] nums, int k) {
       HashMap<Integer,Integer> map = new HashMap<>();
        
        for(int val:nums){
            map.put(val,map.getOrDefault(val,0) + 1);
        }
        
        int ans = 0;
        for(int val:map.keySet()){
            if(k == 0 && map.get(val) > 1){
                ans += 1;
            }
            else if(k != 0 && map.containsKey(val + k)){
                ans += 1;
            }
        }
        return ans;
    }
}

File: 532-k-diff-pairs-in-an-array/532-k-diff-pairs-in-an-array.cpp
class Solution {
public:
    int solve(vector<int> &nums) {
        unordered_map<int,int> mp;
        
        for(auto &e:nums) mp[e]++;
        int ans = 0;
        for(auto &e:mp) {
            if(e.second > 1) ans++;
        }
        return ans;
    }
    int findPairs(vector<int>& nums, int k) {
        sort(nums.begin(),nums.end());
        int n = nums.size();
        if(n <= 1) return 0;
        
        
        if(k == 0) {
            return solve(nums);
        }
        
        unordered_set<int> st;
        
        
        
        st.insert(nums[0]);
        int ans = 0;
        for(int i=1;i<n;i++) {
            if(nums[i] == nums[i-1]) continue;
            
            if(st.find(nums[i] - k) != st.end()) {
                ans += 1;
            }
            st.insert(nums[i]);
        }
    
        return  ans;
    }
};